allennlp/tests/training/multi_task_trainer_test.py

```diff
@@ -0,0 +1,284 @@
+
+    def _read(self, file_path: str):
+        with open(file_path) as data_file:
+            for line in data_file:
+                yield self.text_to_instance(line)
+
+
+class DatasetMingler(Registrable):
+    """
+    Our ``DataIterator`` class expects a single dataset;
+    this is an abstract class for combining multiple datasets into one.
+
+    You could imagine an alternate design where there is a
+    ``MinglingDatasetReader`` that wraps multiple dataset readers,
+    but then somehow you'd have to get it multiple file paths.
+    """
+    def mingle(self, datasets: Dict[str, Iterable[Instance]]) -> Iterable[Instance]:
+        raise NotImplementedError
+
+
+@DatasetMingler.register("round-robin")
+class RoundRobinMingler(DatasetMingler):
+    """
+    Cycle through datasets, ``take_at_time`` instances at a time.
+    """
+    def __init__(self,
+                 dataset_name_field: str = "dataset",
+                 take_at_a_time: int = 1) -> None:
+        self.dataset_name_field = dataset_name_field
+        self.take_at_a_time = take_at_a_time
+
+    def mingle(self, datasets: Dict[str, Iterable[Instance]]) -> Iterable[Instance]:
+        iterators = {name: iter(dataset) for name, dataset in datasets.items()}
+        done: Set[str] = set()
+
+        while iterators.keys() != done:
+            for name, iterator in iterators.items():
+                if name not in done:
+                    try:
+                        for _ in range(self.take_at_a_time):
+                            instance = next(iterator)
+                            instance.fields[self.dataset_name_field] = MetadataField(name)
+                            yield instance
+                    except StopIteration:
+                        done.add(name)
+
+
+@DataIterator.register("homogeneous-batch")
+class HomogeneousBatchIterator(DataIterator):
+    """
+    An iterator that takes instances of various types
+    and yields single-type batches of them. There's a flag
+    to allow mixed-type batches, but at that point you might
+    as well just use ``BasicIterator``?
+    """
+    def __init__(self,
+                 type_field_name: str = "dataset",
+                 allow_mixed_batches: bool = False,
+                 batch_size: int = 32) -> None:
+        super().__init__(batch_size)
+        self.type_field_name = type_field_name
+        self.allow_mixed_batches = allow_mixed_batches
+
+    def _create_batches(self, instances: Iterable[Instance], shuffle: bool) -> Iterable[Batch]:
+        """
+        This method should return one epoch worth of batches.
+        """
+        hoppers: Dict[Any, List[Instance]] = defaultdict(list)
+
+        for instance in instances:
+            # Which hopper do we put this instance in?
+            if self.allow_mixed_batches:
+                instance_type = ""
+            else:
+                instance_type = instance.fields[self.type_field_name].metadata  # type: ignore
```

Comment:
This isn't a real class, so it's not a big deal, but someone might look at this at some point for an example, so I'll leave a comment: you could instead just define `instance_type` to be a string-ified version of `instance.fields.keys()`.  That's really what matters for the model, anyway.  Then you don't need `type_field_name`, or adding a field in `RoundRobinMingler`, or making sure the model can handle that extra field.

Reply:
Fair enough, but that's a lot easier to handle with a `SingleTaskTrainer`, so it's not really clear why you should go to all of this trouble for that case.

Reply:
I'm imagining the case (maybe this is a dumb example) where you have e.g. three different text classification problems, so every instance looks like {"tokens": ..., "label": ..., "dataset": ...}, but they're read from 3 different files, and your model has a different last layer depending on the dataset, does that make sense?